File: pa/pa2/Lex.c
//-----------------------------------------------------------------------------
//  Lex.c
//  Ameer Khan
//  ammukhan
//  CS101 PA2
//  Takes two command line arguments giving the names of an 
//  input file and an output file.
//-----------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "List.h"

#define MAX_LEN 160

int main(int argc, char * argv[]){

   int i;
   int count = 0;
   FILE *in, *out;
   char line[MAX_LEN];

   // check command line for correct number of arguments
   if (argc != 3) {
      printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");

   if (in == NULL) {
      printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if (out == NULL) {
      printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }

   //read each line of input file, then count and print tokens 
   while (fgets(line, MAX_LEN, in) != NULL)  {
      count++;
   }

   //reset file to start
   rewind(in);
   char number[count - 1][MAX_LEN];
   int k = 0;

   while (fgets(line, MAX_LEN, in) != NULL)  {
      strcpy(number[k++], line);
   }


   List list = newList(); //create new list
   append(list, 0); //append 0 to new list

   //insertion sort algorithm
   for (int j = 1; j < count; j++) {
   		i = (j - 1);
   		char *temp = number[j];
   		moveBack(list);
   		while (i >= 0 && strcmp(temp, number[get(list)]) <= 0) {
   			movePrev(list);
   			i--;
   		}
   		if (index(list) >= 0){
   			insertAfter(list, j);
   		} else {
   			prepend(list, j);
   		}
   	}

   	//writes and prints to output file
   	moveFront(list);
   	while (index(list) >= 0) {
   		fprintf(out, "%s", number[get(list)]);
   		moveNext(list);
   	}

   	//closes file and frees list
    fclose(in);
    fclose(out);
    freeList(&list);

    return 0;
}

File: pa/pa1/Lex.java
//-----------------------------------------------------------------------------
//  Lex.java
//  Ameer Khan
//  ammukhan
//  CS101 PA1
//  Takes two command line arguments giving the names of an 
//  input file and an output file.
//-----------------------------------------------------------------------------

import java.io.*;
import java.util.Scanner;
   
class Lex {
   public static void main(String[] args) throws IOException{
      Scanner in = null;
      PrintWriter out = null;
      String line = null;
      String[] token = null;
      List list = new List();
      int lineNumber = 0;
      int i = 0;

      // checks for exactly two arguments otherwise error
      if(args.length != 2){
         System.err.println("Usage: FileIO infile outfile");
         System.exit(1);
      }
      
      in = new Scanner(new File(args[0])); // takes input file
      out = new PrintWriter(new FileWriter(args[1])); // proudces output file

      while (in.hasNextLine()) {
         lineNumber++;
         in.nextLine();
      }
      in.close();
      in = null;


      token = new String[lineNumber];
      in = new Scanner(new File(args[0])); // takes input file
      out = new PrintWriter(new FileWriter(args[1])); // produces output file

      for (i = 0; i < lineNumber; i++) {
         line = in.nextLine()+" "; // add extra space so split works right
         token[i] = line;
      }
      list.append(0);

      // insertion sort algorithm
      for (int j = 1; j < token.length; j++) {
         i = (j - 1);
         String temp = token[j];
         list.moveBack();
         while (i >= 0 && (temp.compareTo(token[list.get()]) <= 0)) {
            list.movePrev();
            i--;
         }
         if (list.index() >= 0) {
            list.insertAfter(j);
         } else {
            list.prepend(j);
         }
      }

      list.moveFront();
      while (list.index() >= 0) {
         out.println(token[list.get()]);
         list.moveNext();
      }

      in.close();
      out.close();
   }
}